node/src/components/linear_chain.rs

```diff
@@ -444,58 +451,30 @@ where
-                    .map(|is_bonded| {
-                        if is_bonded {
-                            Ok((maybe_signatures, fs, is_bonded))
-                        } else {
-                            Err((maybe_signatures, fs))
-                        }
-                    })
-            }
-            .result(
-                |(maybe_signatures, fs, is_bonded)| {
-                    Event::IsBonded(maybe_signatures, fs, is_bonded)
-                },
-                |(maybe_signatures, fs)| Event::IsBondedFutureEra(maybe_signatures, fs),
-            ),
-            Event::IsBondedFutureEra(maybe_signatures, fs) => {
-                match self.latest_block.as_ref() {
-                    // If we don't have any block yet, we cannot determine who is bonded or not.
-                    None => effect_builder
-                        .immediately()
-                        .event(move |_| Event::IsBonded(maybe_signatures, fs, false)),
-                    Some(block) => {
-                        let latest_header = block.header();
-                        let state_root_hash = latest_header.state_root_hash();
-                        // TODO: Use protocol version that is valid for the block's height.
-                        let protocol_version = ProtocolVersion::new(SemVer::V1_0_0);
-                        effect_builder
-                            .is_bonded_in_future_era(
-                                *state_root_hash,
-                                fs.era_id,
-                                protocol_version,
-                                fs.public_key,
-                            )
-                            .map(|res| {
-                                match res {
-                                    // Promote this error to a non-error case.
-                                    // It's not an error that we can't find the era that this
-                                    // FinalitySignature is for.
-                                    Err(error) if error.is_era_validators_missing() => Ok(false),
-                                    _ => res,
-                                }
-                            })
-                            .result(
-                                |is_bonded| Event::IsBonded(maybe_signatures, fs, is_bonded),
-                                |error| {
-                                    error!(%error, "is_bonded_in_future_era returned an error.");
-                                    panic!("couldn't check if validator is bonded")
-                                },
-                            )
-                    }
-                }
+                    .is_bonded_validator(
+                        fs.public_key,
+                        fs.era_id,
+                        activation_era_id,
+                        initial_state_root_hash,
+                        latest_state_root_hash,
+                        protocol_version,
```

Comment:
Shouldn't we be using the protocol version according to the chainspec loader?  We only want to get validators from the auction contract around genesis or after an emergency restart.  Doesn't the chainspec loader contain this information?

Reply:
I think we do, but that might not be entirely correct - we could, in theory, receive a finality signature from before an upgrade, after we upgraded. Then, in order to check that signature, we should be issuing a request with an earlier protocol version (although with the current code it wouldn't be used, anyway - in fact, with my change, we'd not get the validators set at all :grimacing: I should fix that; but after a fix, we'd still be using the set from a key block).

Actually, if we ever need to check the set of validators from before the last upgrade, I think there's an edge case we'd be unable to handle correctly: if there is an upgrade just one era after an emergency restart. Say we had an emergency restart in era `N-1` and an upgrade in era `N`. We are in era `N` after the upgrade and we receive a finality signature for a block in era `N-1`. There is _no way_ we could validate it properly, because being after an upgrade, we have no information about the activation point in era `N-1` and we are unable to get the correct set of validators for that era... :grimacing: